internal/evmreader/output.go

```diff
@@ -21,11 +22,20 @@ func (r *Service) checkForOutputExecution(
 
 	r.Logger.Debug("Checking for new Output Executed Events", "apps", appAddresses)
 
-	for _, app := range apps {
+	mostRecentBlockNumberCallOpts := &bind.CallOpts{
+		BlockNumber: new(big.Int).SetUint64(mostRecentBlockNumber),
+	}
 
-		// Safeguard: Only check blocks starting from the block where the InputBox
-		// contract was deployed as Inputs can be added to that same block
-		lastOutputCheck := max(app.application.LastOutputCheckBlock, app.application.IInputBoxBlock)
+	for _, app := range apps {
+		lastOutputCheck := app.application.LastOutputCheckBlock
+		if lastOutputCheck == 0 { // New application. Find a safe start block to scan for outputs
+			lastOutputCheck = findSafeFirstInputBlockToScan(&app, mostRecentBlockNumber, mostRecentBlockNumberCallOpts) - 1
```

Comment:
Why are we using `findSafeFirstInputBlockToScan` to scan for outputs?
Should we rename it to something more generic?

Reply:
Here are some counterexamples:

- The authority owner's PK got stolen and approved an invalid outputs root
- A majority of the quorum submitted an invalid outputs root
- Only an invalid commitment was submitted to PRT and won by timeout

Since v2 allows for empty epochs, an application could have an invalid outputs root finalized but no inputs received; so it would be possible for an `OutputExecuted` event to be emitted before an `InputAdded` event for that app.

Reply:
Interesting. So the assumption is only valid for a working application.
But may not hold for a broken one.

I am thinking if the node should care about those cases.
Because all bets are off if the optimistic-rollup doesn't have a honest validator.

My gut feeling is that it shouldn't. The application is broken beyond repair.
But we should still detect something is wrong and stop processing this application.

@vfusco have you thought about this case?

Reply:
I agree. If an output is executed before an input is added, the app is broken.
I understand if the node treats this as a "doesn't care" state.